homeassistant/components/sql/sensor.py

```diff
@@ -37,62 +37,80 @@ def redact_credentials(data: str) -> str:
-    _LOGGER.warning(
-        # SQL config flow added in 2022.4 and should be removed in 2022.6
-        "Configuration of the SQL sensor platform in YAML is deprecated and "
-        "will be removed in Home Assistant 2022.6; Your existing configuration "
-        "has been imported into the UI automatically and can be safely removed "
-        "from your configuration.yaml file"
-    )
+    """Set up the SQL sensor from yaml."""
+    if discovery_info is None:
+        return
 
-    default_db_url = DEFAULT_URL.format(
-        hass_config_path=hass.config.path(DEFAULT_DB_FILE)
-    )
+    conf = discovery_info
 
-    for query in config[CONF_QUERIES]:
-        new_config = {
-            CONF_DB_URL: config.get(CONF_DB_URL, default_db_url),
-            CONF_NAME: query[CONF_NAME],
-            CONF_QUERY: query[CONF_QUERY],
-            CONF_UNIT_OF_MEASUREMENT: query.get(CONF_UNIT_OF_MEASUREMENT),
-            CONF_VALUE_TEMPLATE: query.get(CONF_VALUE_TEMPLATE),
-            CONF_COLUMN_NAME: query[CONF_COLUMN_NAME],
-        }
-        hass.async_create_task(
-            hass.config_entries.flow.async_init(
-                DOMAIN,
-                context={"source": SOURCE_IMPORT},
-                data=new_config,
-            )
+    if not (db_url := conf.get(CONF_DB_URL)):
+        db_url = DEFAULT_URL.format(hass_config_path=hass.config.path(DEFAULT_DB_FILE))
+
+    sess: scoped_session | None = None
+    try:
+        engine = sqlalchemy.create_engine(db_url)
+        sessmaker = scoped_session(sessionmaker(bind=engine))
+
+        # Run a dummy query just to test the db_url
+        sess = sessmaker()
+        sess.execute("SELECT 1;")
+
+    except SQLAlchemyError as err:
+        _LOGGER.error(
+            "Couldn't connect using %s DB_URL: %s",
+            redact_credentials(db_url),
+            redact_credentials(str(err)),
         )
+        return
+    finally:
+        if sess:
+            sess.close()
+
+    name: str = conf[CONF_NAME]
+    query_str: str = conf[CONF_QUERY]
+    unit: str | None = conf.get(CONF_UNIT_OF_MEASUREMENT)
+    value_template: Template | None = conf.get(CONF_VALUE_TEMPLATE)
+    column_name: str = conf[CONF_COLUMN_NAME]
+    unique_id: str | None = conf.get(CONF_UNIQUE_ID)
+
+    if value_template is not None:
+        value_template.hass = hass
+
+    # MSSQL uses TOP and not LIMIT
+    if not ("LIMIT" in query_str.upper() or "SELECT TOP" in query_str.upper()):
+        query_str = (
+            query_str.replace("SELECT", "SELECT TOP 1")
+            if "mssql" in db_url
+            else query_str.replace(";", " LIMIT 1;")
+        )
+
+    async_add_entities(
+        [
+            SQLSensor(
+                name,
+                sessmaker,
+                query_str,
+                column_name,
+                unit,
+                value_template,
+                unique_id,
+                True,
+            )
+        ],
+        True,
+    )
```

Comment:
This looks like duplicate or very similar code as with `async_setup_entry`

Can you break this out into a function to make it a bit more DRY?